Add RobotInterface library and a SensorBridge interface class

prashanthr05 opened the pull request:
This PR,
- Adds `RobotInterface` library to the Framework
- Adds `ISensorBridge` interface class to `RobotInterface` library
~~- Adds OpenCV as a dependency for handling `cv::Mat` within the `ISensorBridge` class~~
- Adds a trivial `DummySensorBridge` test implementation to depict the use of the `ISensorBridge` interface

S-Dafarra commented:
> Adds OpenCV as a dependency for handling cv::Mat within the ISensorBridge class

To be honest, I think that ``OpenCV`` may be a too strong dependency for a single type. It should be possible to convert a ``cv::Mat`` type from/to an Eigen type, https://github.com/opencv/opencv/blob/e3d502310f6ae0cad79f3474b9ea2b25705985f9/modules/core/include/opencv2/core/eigen.hpp

What about using Eigen types in the interface then?

prashanthr05 commented:
> To be honest, I think that `OpenCV` may be a too strong dependency for a single type. It should be possible to convert a `cv::Mat` type from/to an Eigen type, https://github.com/opencv/opencv/blob/e3d502310f6ae0cad79f3474b9ea2b25705985f9/modules/core/include/opencv2/core/eigen.hpp
> 
> What about using Eigen types in the interface then?


I thought about it too. If we can depend on the user to make relevant conversions from Eigen to relevant types, then we can keep the design that way.

In that case, would it also make sense to change GenericVectorContainer to Eigen?

prashanthr05 commented:
But then, if we also want point clouds to be handled by this Bridge, then there comes a problem of adding PCL which is quite a heavy library that depends on many other libraries as a dependency. 

So we might have to think about this a bit.

S-Dafarra commented:
> In that case, would it also make sense to change GenericVectorContainer to Eigen?

The ``GenericContainer::Vector`` was meant to have a unique way to deal with vectors that could have also been resized. This was coming in handy when dealing with parameters. Depending on the implementation and on the needs, we may have needed a ``Yarp::Sig::Vector``, or ``iDynTree::VectorDynSize`` for example. This was before moving to Eigen in the main API. So, right now I would suggest using  ``GenericContainer::Vector`` in case you need to deal with different vector implementation at the same time, especially in case you need to resize them (something that is not possible with an ``Eigen::map``).

> But then, if we also want point clouds to be handled by this Bridge, then there comes a problem of adding PCL which is quite a heavy library that depends on many other libraries as a dependency.

Is there a specific type/class we may need to use from this library?

prashanthr05 commented:
> The GenericContainer::Vector was meant to have a unique way to deal with multiple vector implementations that could have also been resized. This was coming in handy when dealing with parameters. Depending on the implementation and on the needs, we may have needed a Yarp::Sig::Vector, or iDynTree::VectorDynSize for example. This was before moving to Eigen in the main API. So, right now I would suggest using GenericContainer::Vector in case you need to deal with different vector implementation at the same time, especially in case you need to resize them (something that is not possible with an Eigen::map).

Agreed. 

> Is there a specific type/class we may need to use from this library?

I am just looking forward into our tasks. For example, for the staircase climbing task, if we need perception in the loop, we could begin to use PCL library extensively for plane segmentation. But from within the class, we would need pcl::PoincloudXYZRGB.

S-Dafarra commented:
> I am just looking forward into our tasks. For example, for the staircase climbing task, if we need perception in the loop, we could begin to use PCL library extensively for plane segmentation. But from within the class, we would need pcl::PoincloudXYZRGB.

I think there might be some conversions available. https://github.com/PointCloudLibrary/pcl/blob/2c8d58a0e6497d91cc50f61460d75b9ebf44429a/common/include/pcl/point_cloud.h#L351

Anyway, the take-home message would be to avoid heavy dependencies only for a type in an interface class. For sure, if we need some additional feature of these libraries, it makes total sense to add them. In principle, I may want to avoid having a certain dependency, paying the price of losing some features. But I am expecting the feature to be related to the dependency I am missing. In this specific case, it would be a pity not to be able to compile the ``RobotInterface`` part if OpenCV is missing. This may complicate things if we would need this to run on the robot head maybe.

prashanthr05 commented:
> Anyway, the take-home message would be to avoid heavy dependencies only for a type in an interface class. For sure, if we need some additional feature of these libraries, it makes total sense to add them. In principle, I may want to avoid having a certain dependency, paying the price of losing some features. But I am expecting the feature to be related to the dependency I am missing. In this specific case, it would be a pity not to be able to compile the RobotInterface part if OpenCV is missing. This may complicate things if we would need this to run on the robot head maybe

I understand. In that case, we can avoid having OpenCV and PCL as a dependency for this interface class and rely on relevant conversions at a higher level. I will replace the image related methods to Eigen types. 
For the PCL, we shall deal with it in the near future.

prashanthr05 commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,296 @@
+
+    /**
+     * Get gyroscopes list
+     */
+    virtual const std::vector<std::string> getGyroscopesList() = 0;
+
+    /**
+     * Get orientation sensors list
+     */
+    virtual const std::vector<std::string> getOrientationSensorsList() = 0;
+
+    /**
+     * Get magnetometers sensors list
+     */
+    virtual const std::vector<std::string> getMagnetometersList() = 0;
+
+    /**
+     * Get 6 axis FT sensors list
+     */
+    virtual const std::vector<std::string> getSixAxisForceTorqueSensorsList() = 0;
+
+    /**
+     * Get 6 axis FT sensors list
+     */
+    virtual const std::vector<std::string> getThreeAxisForceTorqueSensorsList() = 0;
+
+   /**
+     * Get cartesian wrenches list
+     */
+    virtual const std::vector<std::string> getCartesianWrenchesList() = 0;
+
+    /**
+     * Get rgb cameras list
+     */
+    virtual const std::vector<std::string> getRGBCamerasList() = 0;
+
+    /**
+     * Get depth cameras list
+     */
+    virtual const std::vector<std::string> getDepthCamerasList() = 0;
```
Oh yes.. that was the initial intention, but I changed my mind since it would depend on whether the implementation will hold the data structure in their class. So I thought return by value would be better, but I forgot to remove the const.

S-Dafarra replied:
This would be a big waste of memory. I think that, in this case, the possible choices would be:
- return by reference
- have a reference to the vector as an input to the method

prashanthr05 replied:
I think we can go for the latter without having to impose the maintenance of those data structures by the implementation class.

S-Dafarra replied:
As you prefer!

prashanthr05 replied:
But actually, to be extra safe, just not return the reference of a string that "might" be local, I will use the second choice.

S-Dafarra replied:
> But actually, to be extra safe, just not return the reference of a string that "might" be local, I will use the second choice.

I guess you meant ``vector`` and not ``string``, but yeah, I agree there is the risk that in a faulty implementation a reference to a local object is returned.

S-Dafarra replied:
Actually, I realized later that these methods are probably called just once, so even return them by copy is not a big deal. So, feel free to return directly the vector if you believe the interface would be cleaner.

prashanthr05 replied:
Actually I already changed them to be passed by reference and then copied. We can go with this choice, which can also handle failure cases through the boolr eturn.

S-Dafarra replied:
Sounds good!

prashanthr05 replied:
Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/666670f08e102f679fff287bc896cadc7b91de14

prashanthr05 commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,296 @@
+     */
+    virtual bool getJointPosition(const std::string& jointName,
+                                  double& jointPosition,
+                                  double& recvTimeInS) = 0;
```
> Even if it is a little uglier, a possibility could be to input a double pointer instead of a reference, and default it to a nullptr. Then, inside, there should be a check on whether this pointer is null or not. What do you think?

It's not really clear to me the purpose of this suggestion? Could you explain a bit more?

S-Dafarra replied:
I was thinking at something like:
```c++
virtual bool getSomething(SomeType& something,
                                  double* recvTimeInS = nullptr) = 0;
```
In this way, you can call it like
```c++
SomeType output;
getSomething(output);
```
and if you are also interested in the time
```c++
double time;
SomeType output;
getSomething(output, &time);
```

Another possibility could be also to have both the methods with and without ``recvTimeInS``, where the latter simply calls the former with a dummy ``double``, i.e.
```c++
virtual bool getSomething(SomeType& something,
                                  double& recvTimeInS) = 0;

bool getSomething(SomeType& something,
                                  double& recvTimeInS) //Note, this is not virtual
{
    double dummy;
    return getSomething(something, dummy);
}
```
Does it make sense?

prashanthr05 replied:
Ah, ok! I got your point now. I will make the changes corresponding to the former choice.

prashanthr05 replied:
Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/3bbf33a17441c9a8ed86cf5c2528659a2cbb5f5e

S-Dafarra commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,318 @@
+};
+
+/**
+ *  Senosr lists
```
Typo

S-Dafarra commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,318 @@
+/**
+ * Sensor bridge options
+ */
+struct SensorBridgeOptions
```
How are you planning to use these options?

prashanthr05 replied:
It's just like a board of switches to turn on/off the processing during the initialization step. An instance of the SensorBridge will not consider attaching itself to sensor streams that are not enabled.

S-Dafarra replied:
It makes sense but I did not see it used in the interface. What would be your plan?

prashanthr05 replied:
The intention was that the Derived class will hold private instances of `SensorBridgeOptions` and `SensorLists` for internal organization. I shall add a line of documentation on how it can be used.

S-Dafarra replied:
Ok. In any case, my suggestion would be to declare what you use and leave freedom to the user for the rest. If you declare some struct in the base interface, but the base interface does not use them (neither as an attribute nor as an input parameter to a method), then there is a big chance they are never used, or used in a wrong way 😁

prashanthr05 replied:
I wanted to have them within the RobotInterface namespace. Would it be better if I moved it to a new header keeping it separate from the interface?

S-Dafarra replied:
Do they use it in the same way? My main point is that either it is clear from the code how to use these structs, or it may be better to leave the freedom to the user.

A suggestion I may give is to have a function that automatically fills this struct given a pointer to a parameter handler, for example.

prashanthr05 replied:
I can add a protected method in the interface as per your suggestion.

S-Dafarra replied:
Sounds good!

prashanthr05 replied:
I added optional helper methods in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/4a4b818f9bb3ad844993a7a6596c16844e7af942 to avoid the structs from remaining dangling.
I added a line of documentation which will allow the user to also avoid using these structs if they decide to implement their own better ways.

S-Dafarra commented on src/RobotInterface/tests/DummyImplementation/DummySensorBridgeTest.cpp:
```diff
@@ -0,0 +1,55 @@
+#include <BipedalLocomotion/ParametersHandler/IParametersHandler.h>
+#include <BipedalLocomotion/ParametersHandler/StdImplementation.h>
+
+#include "./DummySensorBridge.h"
```
I guess you can remove ``./``. When using ``"``, it already looks in the current folder.

GiulioRomualdi commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,318 @@
+     * @return true/false in case of success/failure
+     */
+    virtual bool getLinearAccelerometerMeasurement(const std::string& accName,
+                                                   GenericContainer::Vector<double>& accMeasurement,
```
Is it a good idea to use the `GenericContainer::Vector` here? In all the other parts of the code, we use Eigen vectors as intefrace of the class. I would avoid having mixing different kinds of types

GiulioRomualdi replied:
For instance here you can use an `Eigen::Ref<Eigen::Vector3d>` 
And the you can pass every different container (std vector iDynTree or yarp) through an `Eigen::Map`

prashanthr05 replied:
We had a discussion about this at https://github.com/dic-iit/bipedal-locomotion-framework/pull/87#issuecomment-682450524.

prashanthr05 replied:
The important point that was raised was,

> So, right now I would suggest using GenericContainer::Vector in case you need to deal with different vector implementation at the same time, especially in case you need to resize them (something that is not possible with an Eigen::map).

@S-Dafarra ,

> I would avoid having mixing different kinds of types

would you have an opinion on this?

S-Dafarra replied:
> @S-Dafarra ,
> 
> > I would avoid having mixing different kinds of types
> 
> would you have an opinion on this?

I would say that it depends on whether ``accMeasurement`` has a predefined size or not. If so, I agree on using Eigen types for consistency.

prashanthr05 replied:
There are a few quantities that have might have a variable size like images, joint states, etc. Otherwise, the rest are all fixed size.

Anyhow, these are going to remain fixed once the user configures the number of joints or the image dimensions.

GiulioRomualdi replied:
> There are a few quantities that have might have a variable size like images, joint states, etc. Otherwise, the rest are all fixed size.

The images I think are stored in matrices right? If yes in this case `GenericContainer::Vector` cannot be used. For the joint state back in time I simply used `iDynTree::VectorDynSize`, using Eigen the code may become
```cpp
bool SensorBridge::getJointPos(Eigen::Ref<Eigen::VectorXd> v);

....

main()
{
  SensorBridge sb;
  iDynTree::VectorDynSize q_iDynTree;
  yarp::Vector q_yarp;
  Eigen::VectorXd q_eigen;
  sb.getJointPos(iDynTree::toEigen(q_iDynTree));
  sb.getJointPos(yarp::toEigen(q_yarp));
  sb.getJointPos(yarp::toEigen(q_eigen));
}
```

GiulioRomualdi replied:
> > There are a few quantities that have might have a variable size like images, joint states, etc. Otherwise, the rest are all fixed size.
> 
> The images I think are stored in matrices right? If yes in this case `GenericContainer::Vector` cannot be used. For the joint state, back in time I simply used `iDynTree::VectorDynSize`, using Eigen the code may
> 
> ```c++
> bool SensorBridge::getJointPos(Eigen::Ref<Eigen::VectorXd> v);
> 
> ....
> 
> main()become
> {
>   SensorBridge sb;
>   iDynTree::VectorDynSize q_iDynTree;
>   yarp::Vector q_yarp;
>   Eigen::VectorXd q_eigen;
>   sb.getJointPos(iDynTree::toEigen(q_iDynTree));
>   sb.getJointPos(yarp::toEigen(q_yarp));
>   sb.getJointPos(q_eigen);
> }
> ```

By the way this holds only if we don't want to resize the vector inside the `getJointPos()` because in this case, terrible bugs may happen 
```cpp

Eigen::VectorXd q_eigen(100);
 
sb.getJointPos(q_eigen.segment<10>(20)); //<----- If getJointPos resize the vectors terrible things may happen here 
```

prashanthr05 replied:
Relevant changes made in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/4a4b818f9bb3ad844993a7a6596c16844e7af942

GiulioRomualdi replied:
Thank you for handling this :heart:

S-Dafarra replied:
So, how is it handled for the elements with variable size?

prashanthr05 replied:
At the moment, we assume for joint states and images, that the size is decided at the configuration and remains fixed, and internal checks can be done at the implementation level. This means that the user must pass a resized argument to these getter methods. Correct me if I am wrong, @GiulioRomualdi!
I shall add a description in the documentation regarding this.

However, we need to understand how to handle this for the elements with variable size, for example, point clouds.

S-Dafarra replied:
Ok for the documentation. For point clouds, are they matrices or vector of points?

prashanthr05 replied:
They are usually handled as a struct asking for maintaining cloud width and height. Then they store a vector of XYZ points or XYZRGB points. 

I hope we will not be causing any overheads by doing multiple conversions by passing these kinds of images or point cloud measurements through this bridge.

S-Dafarra replied:
I think may open one or more issue to discuss on this. In particular, it is not clear:
- when to use ``Eigen`` vectors vs ``GenericContainer`` vs templates
- how to deal with matrices
- how to deal with particular input classes, like point clouds

prashanthr05 replied:
I have added documentation for the joint states and image methods and relevant configuration parameters in the `SensorBridgeOptions` if the implementation might want to use it. https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/979733e1357ffed832c1533a6c6c464adcdae6b8

S-Dafarra commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,349 @@
+     * @param[in] handler  Parameters handler
+     * @param[in] sensorBridgeOptions SensorBridgeOptions to hold the bridge options for streaming sensor measurements
+     */
+    virtual bool populateSensorBridgOptionsFromConfig(std::weak_ptr<IParametersHandler> handler,
```
``Bridge``

prashanthr05 replied:
Fixed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/979733e1357ffed832c1533a6c6c464adcdae6b8

S-Dafarra commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,349 @@
+     * @param[in] sensorBridgeOptions SensorBridgeOptions to hold the bridge options for streaming sensor measurements
+     */
+    virtual bool populateSensorBridgOptionsFromConfig(std::weak_ptr<IParametersHandler> handler,
+                                                      SensorBridgeOptions& sensorBridgeOptions) = 0;
```
If it is not compulsory, simply return true directly in the header. Otherwise, you can already provide a basic implementation that can be overridden by the user, or even not used since it is protected.

prashanthr05 replied:
I refrained myself from doing that to keep this class abstract. If you think it is not necessarily important to remain as a pure virtual class, I shall add the implementation that simply returns true in the header.

S-Dafarra replied:
Yes, in C++ a class is abstract if at least a method is purely virtual, so no big deal.

prashanthr05 replied:
> A class is made abstract by declaring at least one of its functions as a pure virtual function.

Ah, yes! I was thinking conversely on this.

 I will make the suggested changes then (returning true would be easy :D).

Thank you for this suggestion.

prashanthr05 replied:
Done in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/979733e1357ffed832c1533a6c6c464adcdae6b8

GiulioRomualdi commented on src/RobotInterface/include/BipedalLocomotion/RobotInterface/ISensorBridge.h:
```diff
@@ -0,0 +1,373 @@
+
+#include <Eigen/Dense>
+
+#include <BipedalLocomotion/GenericContainer/Vector.h>
```
This one is not required here, right?

prashanthr05 replied:
Fixed in https://github.com/dic-iit/bipedal-locomotion-framework/pull/87/commits/699ea3f560fc27118e23d9aa30f562487e0ef527

GiulioRomualdi commented:
@prashanthr05 let me know if you want to reorganize the commits otherwise I will merge the PR

prashanthr05 commented:
> @prashanthr05 let me know if you want to reorganize the commits otherwise I will merge the PR

Ciao @GiulioRomualdi .. I have squashed and rebased all changes into a single commit. Thanks!
